Auto forwarder implementation

esinarta opened the pull request:
- Re-activated AutoForwarder and Docker UI components 
- Added default value for endpoint picker rather than having one selected automatically
- Implemented auto-forwarding inside AutoForwarder component
- Created new AutoForwarderWebhookCreated type for GraphQL subscription

johnnyoshika committed:
Revert "Temporarily hide <Docker /> and <AutoForwarder />"

This reverts commit 3c3bb91a4fcd912175af1515d4aca69ab9245603.

esinarta committed:
re-enabled autoforward dock

esinarta committed:
AutoForwarder subscribed to webhookCreated, issue with same webhook being forwarded over and over

esinarta committed:
fixed constant autoforwarding issue

esinarta committed:
added default value to autoforward endpoint selection

esinarta committed:
renamed WebhookCreated subscription to prevent duplicate, refactored unnecessary array, checked for undefined

esinarta committed:
removed unused imports, tidied up spacing

esinarta committed:
changed AutoForward play/recording icons to be consistent size

johnnyoshika commented on src/routes/Viewport/AutoForwarder/index.tsx:
```diff
@@ -22,6 +36,18 @@ const AutoForwarder = ({ docked }: { docked: Boolean }) => {
   const [running, setRunning] = useState(false);
   const { forwardUrls } = useForwardUrls(endpointId);
   const [url, setUrl] = useState('');
+  const { forwardWebhook } = useForwarder(endpointId);
```
When endpointId is empty (i.e. ''), `useForwardUrls(endpointId)` is causing a GraphQL error, as it tries to fetch forward URLs for an endpoint that's empty.  Here is the request that's being sent:

![image](https://user-images.githubusercontent.com/504505/110995383-25741c00-832f-11eb-8f46-15b425e2284f.png)

This results in this response from our GraphQL server:

![image](https://user-images.githubusercontent.com/504505/110995616-826fd200-832f-11eb-879e-0aefa309cb44.png)

Also, every time I change the endpoint from the dropdown, the same request is being sent with an empty endpointId. You can test this by putting a breakpoint here, as that code will be called on every GraphQL error: https://github.com/webhook-buddy/webhook-buddy-client/blob/master/src/index.tsx#L72

You can also see this error being printed to the console:
![image](https://user-images.githubusercontent.com/504505/110995990-19d52500-8330-11eb-8a12-ffc9569f32ee.png)

To fix this, I suggest you extract this into a separate component, and only load it when an appropriate endpoint is selected:
```
        <label>Auto-forward to</label>
        <Autosuggest
          type="url"
          placeholder="Forward to URL (e.g. http://localhost:8000/send-webhook-here)"
          userInput={url}
          setUserInput={setUrl}
          suggestions={forwardUrls}
          position={AutosuggestPositionEnum.Up}
          disabled={running}
        />
      </div>
```

You may also want to extract the endpoint select list to a separate component as well to make it consistent.

johnnyoshika commented:
@esinarta Before you make further changes, can you rebase on top of this main repo's master branch? It looks like your branch hasn't picked up the latest changes. If you haven't set up this repo as a remote repo, you can do something like this:

```
git remote add source https://github.com/webhook-buddy/webhook-buddy-client.git
git fetch source
git rebase source/master
git push -f
```

esinarta commented:
> @esinarta Before you make further changes, can you rebase on top of this main repo's master branch? It looks like your branch hasn't picked up the latest changes. If you haven't set up this repo as a remote repo, you can do something like this:
> 
> ```
> git remote add source https://github.com/webhook-buddy/webhook-buddy-client.git
> git fetch source
> git rebase source/master
> git push -f
> ```

@johnnyoshika Okay, I just followed these steps to rebase.

esinarta committed:
Move AutoForward endpoint dropdown and suggestion list into separate components

BuzzBumble commented:
@johnnyoshika  

> Also, every time I change the endpoint from the dropdown, the same request is being sent with an empty endpointId. 

We couldn't seem to recreate this bug (maybe we fixed it inadvertently?)

> When endpointId is empty (i.e. ''), useForwardUrls(endpointId) is causing a GraphQL error, as it tries to fetch forward URLs for an endpoint that's empty.

This is now fixed, since the AutoSuggest is now in its own conditionally-rendered component and the useForwardURLs hook is being used within that component.

The last concern is that when a webhook is auto-forwarded, the profiler is noticing 10+ re-renders of the AutoForwarder's child components (namely `AutoForwardSuggest`). Still looking into this.

johnnyoshika commented:
> The last concern is that when a webhook is auto-forwarded, the profiler is noticing 10+ re-renders of the AutoForwarder's child components (namely `AutoForwardSuggest`). Still looking into this.

Thanks @BuzzBumble. Would you like me to wait until the rendering issues are resolved or proceed with the review? We can also review the render problem together tomorrow.

BuzzBumble committed:
Moved all individual queries into schema/queries;

BuzzBumble committed:
Regenerated types for queries in /schema;

BuzzBumble committed:
Changed imports to use schema/types;

BuzzBumble committed:
Removed component-specific single use types;

BuzzBumble committed:
Removed FORWARD_FRAGMENT import from schema/queries;

esinarta committed:
Move addForwardUrl from useForwarder into its own hook (useAddForwardUrl) which resolved FORBIDDEN Error

esinarta committed:
Merge branch 'hook_queries' of https://github.com/webhook-buddies/webhook-buddy-client into hook_queries

johnnyoshika commented:
A couple of issues:

* I'm still seeing the `FORBIDDEN error` when I switch endpoints due to the empty `endpointId` being passed to `forwardUrls` query:

![image](https://user-images.githubusercontent.com/504505/113432237-dbacad80-9391-11eb-8740-017b5cf5f042.png)

* I think you can delete everything inside `src/hooks/types`. Our types generation tool doesn't delete existing types. Try deleting that folder and then re-run the type generation. If the deleted ones don't get re-recreated, they were safe to delete.

johnnyoshika commented on src/routes/Viewport/AutoForwarder/index.tsx:
```diff
@@ -1,14 +1,13 @@
-
+import useForwarder from 'hooks/useForwarder';
 import styles from './styles.module.css';
+import { WebhookCreated } from 'hooks/types/WebhookCreated';
```
This is referencing types inside `hooks/types` instead of `schema/types`

esinarta committed:
Merge branch 'hook_queries' into auto-forwarder

esinarta committed:
Delete src/hooks/types/ and re-run types-generate, chang WebhookCreated import to reference schema/types instead of hooks/types

esinarta commented:
I deleted the hooks/types folder and re-ran types-generate and none of the types were re-created. I also changed the WebhookCreated import to reference schema/types.

johnnyoshika commented:
Thanks for the PR. Well done, guys!
